src/tests/base.h

```
@@ -62,9 +98,37 @@ class TestBase
     void storeConfig(ConfigNode& config);
 
     /**
-     * Run the test and return the result
+     * Run the test.
      */
-    virtual Result run() = 0;
+    virtual void run() = 0;
+
+    /**
+     * Get the result of the test just run.
+     */
+    Result getResult() { return _result; };
+
+    template <typename T>
+    void expectEq(const T& a, const T& b, const std::string& a_str, const std::string& b_str,
+                  const std::string& file, int line)
+    {
+        if (a == b) {
+            return;
+        }
+
+        std::cout << "Check at " << extractFilename(file) << ":" << line << " failed" << std::endl;
+        std::cout << a_str << " != " << b_str << std::endl;
+
+        auto maybe_streamable = MaybeStreamable<T>();
+
+        if (maybe_streamable.isStreamable()) {
+            std::cout << maybe_streamable.print(a) << " != " << maybe_streamable.print(b)
+                      << std::endl;
+        }
+
+        _result = Result::Failed;
```

I'd rather throw an exception here. Continue running a failed test does not make much sense to me.

---

It depends on the check. That's why usually there are `expects` and `asserts`. expects should just take note, and asserts should stop.

---

Fine, then we also need an assert.

---

Ok.

---

Implemented.